### PR TITLE
Enable more rules by default

### LINKED ISSUE #2838: Enable more rules by default
The same idea as #1911 9 months later :P. All the rules that we add, we add them disabled by default. So we need to enable by default at some point. Any way we should be carefull here as artur say in the linked issue:

> No "highly subjective" rules should be enabled

I extracted a list of rules that I have enabled by default in my projects and I think that they are "safe":

- VarCouldBeVal
- UseRequire
- UseCheckOrError
- UnusedPrivateMember
- UnnecessaryLet
- UnnecessaryApply
- UnderscoresInNumericLiterals
- SerialVersionUIDInSerializableClass
- RedundantVisibilityModifierRule
- UnsafeCast
- UnnecessarySafeCall
- UnnecessaryNotNullOperator
- ImplicitDefaultLocale
- IgnoredReturnValue
- InvalidPackageDeclaration
- ThrowingExceptionsWithoutMessageOrCause
- SwallowedException
- RethrowCaughtException
- ExceptionRaisedInUnexpectedLocation

I'm not saying that we should enable all of these nor any of the others. This is just a first draft so we have a starting point to decide.

What do you think about these ones? Should we add some other?

### PR DESCRIPTION
**Pull request by schalkms**
This is a first draft for enabling more detekt rules by default.
This way users can benefit from a good default config.

Closes #2838

### CODE REVIEW

File: detekt-core/src/main/resources/default-detekt-config.yml [hunk 1]
```diff
@@ -468,11 +468,11 @@ potential-bugs:
   HasPlatformType:
     active: false
   IgnoredReturnValue:
-    active: false
+    active: true
```
Comment (cortinico) [inline detekt-core/src/main/resources/default-detekt-config.yml:471]: I'm unsure about this. We had several false positives for this rules, specifically from RxJava users.

**Review by cortinico — COMMENTED** [review]
We should also recall to annotate with `@active since ...`

File: detekt-core/src/main/resources/default-detekt-config.yml (hunk 1 above)
Comment (arturbosch) [reply detekt-core/src/main/resources/default-detekt-config.yml:471]: Agree here

**Comment by arturbosch** [comment]
The added rules look good to me :+1:
Lets wait for 1.16.0 for this PR, okay?

We have also some more rules enabled in our analysis which are not active by default:
https://github.com/detekt/detekt/issues/2838#issuecomment-726365903

**Comment by schalkms** [comment]
> Lets wait for 1.16.0 for this PR, okay?

That's okay for me. I'll also check the ones mentioned in your last comment.

**Review by BraisGabin — COMMENTED** [review]
As @cortinico said we need to add the `@active`. Then we could just regenerate the file and the values should be changed automatically.

**Comment by schalkms** [comment]
> we need to add the `@active`

Yep, this is the 2nd step. The 1st step should be the draft of the activated rules.
I’ll add the tag to the rules in the next few days. Fortunately, I had the pleasure to co-create and shape the detekt-verify as well as the -generator back then. 🙂

> typos

Regarding the typos, it’s probably not a good idea to edit the config on the go with a smartphone web browser. Well, we luckily have a good CI system and carefully observing reviewers for those reasons.

Cheers and have a nice weekend,
Markus

**Comment by BraisGabin** [comment]
What do you think about enable the coroutine ones? I think that all are quite safe. The more controversial is `GlobalCoroutineUsage` and the documentation of Kotlin always say "please don't use Global" so it should be safe.

So my list to be enabled is:

- [ ] `GlobalCoroutineUsage`
- [ ] `RedundantSuspendModifier`
- [ ] `SuspendFunWithFlowReturnType`
- [ ] `InvalidPackageDeclaration`
- [ ] `HasPlatformType`
- [ ] `MapGetWithNotNullAssertionOperator`
- [ ] `ForbiddenMethodCall`
- [ ] `ForbiddenVoid`
- [ ] `UnderscoresInNumericLiterals`
- [ ] `UnnecessaryAnnotationUseSiteTarget`
- [ ] `UnnecessaryLet`
- [ ] `UseArrayLiteralsInAnnotations`
- [ ] `UseCheckOrError`
- [ ] `UseIfEmptyOrIfBlank`
- [ ] `UseRequire`
- [ ] `UseRequireNotNull` We should probably merge this one with `UseRequire`. `UseRequire` already handles `check` and `checkNotNull` so there's not a good reason to have this one splitted

Not so sure:

- [ ] `MandatoryBracesIfStatements`
- [ ] `MandatoryBracesLoops`
- [ ] `NullableToStringCall` 🤔 this one is too new but this issue creates bugs so often...

**Comment by schalkms** [comment]
Thanks for your work, @BraisGabin.
I had reasons for not enabling all mentioned rules. Hence, I'm not okay with enabling the following rules.

- `UnderscoresInNumericLiterals` => this one is really a matter of choice. I don't think all agree with using underscores for numeric literals. We shouldn't enable this by default.
- `UnnecessaryLet` => this rule is not battle-tested IMHO. So there might be many false-positives.
- `UseCheckOrError`, `UseRequire` & `UseRequireNotNull`=> we don't even use this throughout detekt's code base consistently. It's a matter of taste which style you use.
- The rules in the **not sure list** should be disabled.

**Comment by BraisGabin** [comment]
> Thanks for your work, @BraisGabin.
> I had reasons for not enabling all mentioned rules. Hence, I'm not okay with enabling the following rules.
> 
> - `UnderscoresInNumericLiterals` => this one is really a matter of choice. I don't think all agree with using underscores for numeric literals. We shouldn't enable this by default.
> - `UnnecessaryLet` => this rule is not battle-tested IMHO. So there might be many false-positives.
> - `UseCheckOrError`, `UseRequire` & `UseRequireNotNull`=> we don't even use this throughout detekt's code base consistently. It's a matter of taste which style you use.
> - The rules in the **not sure list** should be disabled.

I agree with you. We should not enable the rules you list.

**Comment by cortinico** [comment]
> * `UnnecessaryLet` => this rule is not battle-tested IMHO. So there might be many false-positives.

Agree on this

> * ForbiddenMethodCall

I think this should not be enabled as it's a rule users have to configure, with the list of Methods they want to restrict.

**Comment by BraisGabin** [comment]
> I think this should not be enabled as it's a rule users have to configure, with the list of Methods they want to restrict.

We have two default values there for println. And I think that we should forbid that for default.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/SerialVersionUIDInSerializableClass.kt [hunk 2]
```diff
@@ -37,6 +37,8 @@ import org.jetbrains.kotlin.psi.KtProperty
  *     }
  * }
  * </compliant>
+ *
+ * @active since v1.16.0
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/SerialVersionUIDInSerializableClass.kt:41]: Note that Android context, `Parcelable` is preferred over `Serializable`. This rule is almost a no-op for Android code.

File: detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/PrintStackTrace.kt [hunk 3]
```diff
@@ -43,6 +43,8 @@ import org.jetbrains.kotlin.psi.psiUtil.getReceiverExpression
  *     }
  * }
  * </compliant>
+ *
+ * @active since v1.16.0
```
[the remark this answers is not in the document]
Comment (BraisGabin) [reply detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/PrintStackTrace.kt:47]: Every rule should have a descripction so any user can know why he/she should enable/disable it. If you think that in this case (or any other) the description is not good enough you can create an issue. Or, even better, if you know how to improve it, create a PR.

But I don't think that we should add any documentation about why a rule is or is not active by default.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/SerialVersionUIDInSerializableClass.kt (hunk 2 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/SerialVersionUIDInSerializableClass.kt:41]: I agree that in Android we should not use Serializable, but, if we have, we should use the SerialVersionUID. So I think that it's safe to have this enabled.

We talked some times about an Android rule set. A good rule for there would be to flag any use of `Serializable`.

Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/SerialVersionUIDInSerializableClass.kt:41]: I just create a discussion to talk about it: #3303

**Review by arturbosch — APPROVED** [review]
@chao2zhang kudos for resolving the conflicts +1

**Comment by schalkms** [comment]
@chao2zhang thanks for pushing this over the finish line.

**Pull request merged**